File: sdk/typescript/src/envelope.ts
/**
 * AgentAuth Agentic Identity Envelope — TypeScript implementation.
 *
 * The AgenticEnvelope is the core authorization container for AI agents.
 * It is a signed JWT carrying the agent's identity, declared intent, and
 * tool scope, and must be passed with every MCP tool call.
 */

/** The claims inside a signed envelope JWT. */
export interface EnvelopeClaims {
  /** SPIFFE ID of the entity that spawned this agent. */
  delegator_id: string;
  /** SPIFFE SVID of the executing agent. */
  agent_id: string;
  /** Unique identifier for this task execution. */
  task_id: string;
  /** Organization/tenant scope. */
  tenant_id: string;
  /** Session scope grouping related envelopes. */
  session_id: string;
  /** Human-readable task description. */
  declared_intent: string;
  /** The tools this agent is authorized to use. */
  tool_scope: Array<{
    tool: string;
    operations: string[];
    constraints?: Record<string, string>;
  }>;
  /** Reference to recorded user consent. */
  consent_ref?: string;
  /** Signed parent envelopes forming the delegation chain. */
  delegation_chain?: string[];
  /** JWT issued-at timestamp. */
  iat: number;
  /** JWT expiry timestamp. */
  exp: number;
  /** JWT unique identifier for revocation. */
  jti: string;
  /** Envelope format version. */
  ver: string;
  /** JWT issuer. */
  iss?: string;
  /** JWT subject (= agent_id). */
  sub?: string;
}

/** A single tool permission entry. */
export class ToolPermission {
  constructor(
    /** MCP tool name or glob pattern. */
    public readonly tool: string,
    /** Allowed operations. */
    public readonly operations: string[],
    /** Optional constraints. */
    public readonly constraints: Record<string, string> = {},
  ) {}

  /**
   * Check if this permission allows the given tool and operation.
   */
  allows(tool: string, operation: string): boolean {
    if (!matchesPattern(this.tool, tool)) return false;
    return this.operations.some((op) => op === "*" || op === operation);
  }

  toJSON() {
    const obj: Record<string, unknown> = {
      tool: this.tool,
      operations: this.operations,
    };
    if (Object.keys(this.constraints).length > 0) {
      obj.constraints = this.constraints;
    }
    return obj;
  }
}

/**
 * The Agentic Identity Envelope — the core authorization container for AI agents.
 *
 * Parse and verify an envelope from the X-Agent-Envelope header:
 *
 * @example
 * ```typescript
 * import { EnvelopeVerifier } from "@agentauth/sdk";
 *
 * const verifier = new EnvelopeVerifier({ publicKeyPem: PUBLIC_KEY_PEM });
 * const envelope = await verifier.verify(request.headers["x-agent-envelope"]);
 *
 * if (!envelope.hasToolPermission("database.query", "read")) {
 *   throw new Error("Agent not permitted to query database");
 * }
 * ```
 */
export class AgenticEnvelope {
  /** SPIFFE ID of the entity that spawned this agent. */
  readonly delegatorId: string;
  /** SPIFFE SVID of the executing agent. */
  readonly agentId: string;
  /** Unique task identifier. */
  readonly taskId: string;
  /** Organization/tenant scope. */
  readonly tenantId: string;
  /** Session scope. */
  readonly sessionId: string;
  /** Human-readable description of what the agent will do. */
  readonly declaredIntent: string;
  /** The tools this agent is authorized to use. */
  readonly toolScope: ToolPermission[];
  /** Reference to user consent record. */
  readonly consentRef: string;
  /** Signed parent envelopes forming the delegation chain. */
  readonly delegationChain: string[];
  /** Issuance Unix timestamp. */
  readonly issuedAt: number;
  /** Expiry Unix timestamp. */
  readonly expiresAt: number;
  /** Unique JWT ID for revocation. */
  readonly jti: string;
  /** Envelope format version. */
  readonly version: string;

  constructor(claims: EnvelopeClaims) {
    this.delegatorId = claims.delegator_id ?? "";
    this.agentId = claims.agent_id ?? claims.sub ?? "";
    this.taskId = claims.task_id ?? "";
    this.tenantId = claims.tenant_id ?? "";
    this.sessionId = claims.session_id ?? "";
    this.declaredIntent = claims.declared_intent ?? "";
    this.toolScope = (claims.tool_scope ?? []).map(
      (ts) => new ToolPermission(ts.tool, ts.operations, ts.constraints ?? {}),
    );
    this.consentRef = claims.consent_ref ?? "";
    this.delegationChain = claims.delegation_chain ?? [];
    this.issuedAt = claims.iat;
    this.expiresAt = claims.exp;
    this.jti = claims.jti ?? "";
    this.version = claims.ver ?? "1";
  }

  /** Returns true if the envelope has passed its expiration time. */
  get isExpired(): boolean {
    return Date.now() / 1000 > this.expiresAt;
  }

  /** Returns the number of delegation hops in this envelope. */
  get delegationDepth(): number {
    return this.delegationChain.length;
  }

  /**
   * Check whether this envelope grants the specified operation on the given tool.
   *
   * @param tool - Tool name (e.g. "filesystem.read", "github.issues.create")
   * @param operation - Operation (e.g. "read", "write", "delete")
   */
  hasToolPermission(tool: string, operation: string): boolean {
    return this.toolScope.some((perm) => perm.allows(tool, operation));
  }

  /** Returns a plain object representation of the envelope. */
  toJSON(): Record<string, unknown> {
    return {
      delegator_id: this.delegatorId,
      agent_id: this.agentId,
      task_id: this.taskId,
      tenant_id: this.tenantId,
      session_id: this.sessionId,
      declared_intent: this.declaredIntent,
      tool_scope: this.toolScope.map((ts) => ts.toJSON()),
      consent_ref: this.consentRef,
      delegation_chain: this.delegationChain,
      iat: this.issuedAt,
      exp: this.expiresAt,
      jti: this.jti,
      ver: this.version,
    };
  }
}

/** Options for EnvelopeVerifier. */
export interface EnvelopeVerifierOptions {
  /** ECDSA P-256 public key in PEM format. Used for offline verification. */
  publicKeyPem?: string;
  /** Expected JWT issuer. Default "agentauth". */
  issuer?: string;
}

/**
 * Verifies signed Agentic Identity Envelope JWTs.
 *
 * Use this in your MCP server or A2A endpoint to verify incoming envelopes
 * without calling the AgentAuth server (offline verification).
 *
 * @example
 * ```typescript
 * import { EnvelopeVerifier } from "@agentauth/sdk";
 *
 * const verifier = new EnvelopeVerifier({ publicKeyPem: process.env.AGENTAUTH_PUBLIC_KEY });
 *
 * // In your Express/Hono/Fastify middleware:
 * const envelope = await verifier.verify(req.headers["x-agent-envelope"]);
 * if (!envelope.hasToolPermission("github.issues", "write")) {
 *   res.status(403).json({ error: "insufficient_scope" });
 *   return;
 * }
 * ```
 */
export class EnvelopeVerifier {
  private readonly publicKeyPem?: string;
  private readonly issuer: string;

  constructor(options: EnvelopeVerifierOptions = {}) {
    this.publicKeyPem = options.publicKeyPem;
    this.issuer = options.issuer ?? "agentauth";
  }

  /**
   * Verify a signed envelope JWT and return the parsed envelope.
   *
   * @param token - The signed JWT string from the X-Agent-Envelope header.
   * @throws Error if the token is invalid, expired, or has a bad signature.
   */
  async verify(token: string): Promise<AgenticEnvelope> {
    if (!token) {
      throw new Error("envelope token is empty");
    }

    if (this.publicKeyPem) {
      return this.verifyWithKey(token);
    }

    // Fallback: decode without verification (development only)
    console.warn(
      "[AgentAuth] SECURITY WARNING: envelope verification is disabled " +
        "(no publicKeyPem configured). Do not use in production.",
    );
    return this.decodeUnverified(token);
  }

  private async verifyWithKey(token: string): Promise<AgenticEnvelope> {
    // Use Web Crypto API (available in Node.js 18+, browsers, Deno, Bun)
    try {
      const parts = token.split(".");
      if (parts.length !== 3) {
        throw new Error("malformed JWT: expected 3 parts");
      }

      const [headerB64, payloadB64, signatureB64] = parts;

      // Decode header
      const headerJson = base64UrlDecode(headerB64);
      const header = JSON.parse(headerJson) as { alg?: string; typ?: string };

      if (header.alg !== "ES256") {
        throw new Error(`unsupported algorithm: ${header.alg}`);
      }

      // Import the public key
      const pemBody = this.publicKeyPem!
        .replace(/-----BEGIN PUBLIC KEY-----/, "")
        .replace(/-----END PUBLIC KEY-----/, "")
        .replace(/\s+/g, "");
      const keyBytes = Uint8Array.from(atob(pemBody), (c) => c.charCodeAt(0));
      const cryptoKey = await crypto.subtle.importKey(
        "spki",
        keyBytes,
        { name: "ECDSA", namedCurve: "P-256" },
        false,
        ["verify"],
      );

      // Verify signature
      const signingInput = `${headerB64}.${payloadB64}`;
      const signatureBytes = Uint8Array.from(atob(base64UrlToBase64(signatureB64)), (c) =>
        c.charCodeAt(0),
      );
      const dataBytes = new TextEncoder().encode(signingInput);

      // Convert DER/compact signature to Web Crypto format
      const valid = await crypto.subtle.verify(
        { name: "ECDSA", hash: { name: "SHA-256" } },
        cryptoKey,
        signatureBytes,
        dataBytes,
      );

      if (!valid) {
        throw new Error("envelope signature verification failed");
      }

      // Parse claims
      const payloadJson = base64UrlDecode(payloadB64);
      const claims = JSON.parse(payloadJson) as EnvelopeClaims;

      // Validate expiry
      if (Date.now() / 1000 > claims.exp) {
        throw new Error("envelope token has expired");
      }

      // Validate issuer
      if (claims.iss && claims.iss !== this.issuer) {
        throw new Error(`unexpected issuer: ${claims.iss}`);
      }

      return new AgenticEnvelope(claims);
    } catch (err) {
      if (err instanceof Error && err.message.startsWith("envelope")) {
        throw err;
      }
      throw new Error(`envelope verification failed: ${err}`);
    }
  }

  private decodeUnverified(token: string): AgenticEnvelope {
    const parts = token.split(".");
    if (parts.length !== 3) {
      throw new Error("malformed JWT: expected 3 parts");
    }
    const payloadJson = base64UrlDecode(parts[1]);
    const claims = JSON.parse(payloadJson) as EnvelopeClaims;
    return new AgenticEnvelope(claims);
  }
}

// ---- Utility functions ----

function base64UrlDecode(b64url: string): string {
  const b64 = base64UrlToBase64(b64url);
  return atob(b64);
}

function base64UrlToBase64(b64url: string): string {
  return b64url.replace(/-/g, "+").replace(/_/g, "/") + "==".slice(0, (4 - (b64url.length % 4)) % 4);
}

function matchesPattern(pattern: string, value: string): boolean {
  if (pattern === "*" || pattern === value) return true;
  if (pattern.endsWith(".*")) {
    const prefix = pattern.slice(0, -2);
    return value === prefix || value.startsWith(prefix + ".");
  }
  if (pattern.endsWith("*")) {
    return value.startsWith(pattern.slice(0, -1));
  }
  return false;
}

File: sdk/python/agentauth/envelope.py
"""
AgentAuth envelope module — create, sign, and verify Agentic Identity Envelopes.

For most use cases, use AgentAuthClient.create_envelope() instead of this module
directly. This module is for low-level envelope operations when you need to
verify envelopes without calling the server.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any, Optional

# Optional JWT library — use PyJWT if available
try:
    import jwt as pyjwt  # type: ignore[import]
    _HAS_PYJWT = True
except ImportError:
    _HAS_PYJWT = False


@dataclass
class ToolPermission:
    """Defines the tools and operations an agent is permitted to use."""

    #: The MCP tool name or glob pattern.
    #: Examples: "filesystem.read", "github.issues.*", "postgres://db/*"
    tool: str

    #: The allowed operations. Use ["*"] for all operations.
    operations: list[str]

    #: Optional additional constraints (e.g. {"max_rows": "1000"})
    constraints: dict[str, str] = field(default_factory=dict)

    def allows(self, tool: str, operation: str) -> bool:
        """Check if this permission allows the given tool and operation."""
        if not _matches_pattern(self.tool, tool):
            return False
        return any(op == "*" or op == operation for op in self.operations)

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"tool": self.tool, "operations": self.operations}
        if self.constraints:
            d["constraints"] = self.constraints
        return d

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ToolPermission":
        return cls(
            tool=data["tool"],
            operations=data.get("operations", []),
            constraints=data.get("constraints") or {},
        )


@dataclass
class AgenticEnvelope:
    """
    The Agentic Identity Envelope carries the complete authorization context
    for an AI agent — who it is, who delegated it, what it may do, and for how long.

    This is the core data structure of AgentAuth. Every MCP tool call and
    A2A delegation must carry a signed envelope.
    """

    #: SPIFFE ID of the entity that spawned this agent.
    delegator_id: str

    #: SPIFFE SVID of the executing agent.
    agent_id: str

    #: Unique identifier for this task execution.
    task_id: str

    #: Organization/tenant scope.
    tenant_id: str

    #: Session scope grouping related envelopes.
    session_id: str

    #: Human-readable description of what the agent will do.
    declared_intent: str

    #: The tools this agent is authorized to use.
    tool_scope: list[ToolPermission]

    #: Reference to the recorded user consent.
    consent_ref: str

    #: Signed parent envelopes forming the delegation chain.
    delegation_chain: list[str]

    #: Unix timestamp of issuance.
    issued_at: int

    #: Unix timestamp of expiry.
    expires_at: int

    #: Unique JWT ID for revocation.
    jti: str

    #: Envelope format version.
    version: str = "1"

    @property
    def is_expired(self) -> bool:
        """Returns True if the envelope has passed its expiration time."""
        return time.time() > self.expires_at

    @property
    def delegation_depth(self) -> int:
        """Returns the number of delegation hops in this envelope."""
        return len(self.delegation_chain)

    def has_tool_permission(self, tool: str, operation: str) -> bool:
        """
        Check whether this envelope grants the specified operation on the given tool.

        :param tool: The tool name (e.g. "filesystem.read", "github.issues.create").
        :param operation: The operation (e.g. "read", "write", "delete").
        :returns: True if the tool/operation is permitted.
        """
        return any(perm.allows(tool, operation) for perm in self.tool_scope)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "AgenticEnvelope":
        """Deserialize an AgenticEnvelope from a claims dictionary."""
        return cls(
            delegator_id=data.get("delegator_id", ""),
            agent_id=data.get("agent_id", data.get("sub", "")),
            task_id=data.get("task_id", ""),
            tenant_id=data.get("tenant_id", ""),
            session_id=data.get("session_id", ""),
            declared_intent=data.get("declared_intent", ""),
            tool_scope=[
                ToolPermission.from_dict(ts)
                for ts in data.get("tool_scope", [])
            ],
            consent_ref=data.get("consent_ref", ""),
            delegation_chain=data.get("delegation_chain") or [],
            issued_at=data.get("iat", 0),
            expires_at=data.get("exp", 0),
            jti=data.get("jti", ""),
            version=data.get("ver", "1"),
        )

    def to_dict(self) -> dict[str, Any]:
        """Serialize the envelope to a dictionary."""
        return {
            "delegator_id": self.delegator_id,
            "agent_id": self.agent_id,
            "task_id": self.task_id,
            "tenant_id": self.tenant_id,
            "session_id": self.session_id,
            "declared_intent": self.declared_intent,
            "tool_scope": [ts.to_dict() for ts in self.tool_scope],
            "consent_ref": self.consent_ref,
            "delegation_chain": self.delegation_chain,
            "iat": self.issued_at,
            "exp": self.expires_at,
            "jti": self.jti,
            "ver": self.version,
        }


class EnvelopeVerifier:
    """
    Verifies signed Agentic Identity Envelope JWTs using the AgentAuth public key.

    Use this in your MCP server or A2A endpoint to verify incoming envelopes
    without calling the AgentAuth server (offline verification).

    Example::

        from agentauth import EnvelopeVerifier

        verifier = EnvelopeVerifier(public_key_pem=PUBLIC_KEY_PEM)

        # In your request handler:
        envelope = verifier.verify(request.headers["X-Agent-Envelope"])
        if not envelope.has_tool_permission("filesystem.read", "read"):
            raise PermissionError("Agent not permitted to read filesystem")
    """

    def __init__(self, public_key_pem: str, issuer: str = "agentauth") -> None:
        """
        :param public_key_pem: The ECDSA P-256 public key in PEM format.
        :param issuer: The expected JWT issuer claim value.
        """
        self.public_key_pem = public_key_pem
        self.issuer = issuer
        self._key: Any = None

        if _HAS_PYJWT:
            from cryptography.hazmat.primitives.serialization import load_pem_public_key  # type: ignore[import]
            try:
                self._key = load_pem_public_key(public_key_pem.encode())
            except Exception as e:
                raise ValueError(f"Invalid public key PEM: {e}") from e

    def verify(self, token: str) -> AgenticEnvelope:
        """
        Verify a signed envelope JWT and return the parsed envelope.

        :param token: The signed JWT string from the X-Agent-Envelope header.
        :returns: Parsed AgenticEnvelope with verified claims.
        :raises ValueError: If the token is invalid, expired, or has bad signature.
        """
        if not token:
            raise ValueError("envelope token is empty")

        if _HAS_PYJWT and self._key is not None:
            try:
                claims = pyjwt.decode(
                    token,
                    self._key,
                    algorithms=["ES256"],
                    options={"require": ["exp", "iat", "iss", "sub"]},
                    issuer=self.issuer,
                )
                return AgenticEnvelope.from_dict(claims)
            except pyjwt.ExpiredSignatureError:
                raise ValueError("envelope token has expired")
            except pyjwt.InvalidTokenError as e:
                raise ValueError(f"invalid envelope token: {e}")
        else:
            # Fallback: decode without signature verification (dev mode only)
            # NEVER use this in production
            return self._decode_unverified(token)

    def _decode_unverified(self, token: str) -> AgenticEnvelope:
        """Decode without signature verification — for development only."""
        import base64
        parts = token.split(".")
        if len(parts) != 3:
            raise ValueError("malformed JWT: expected 3 parts")

        # Add padding for base64url decoding
        payload_b64 = parts[1] + "==" * (4 - len(parts[1]) % 4)
        payload_bytes = base64.urlsafe_b64decode(payload_b64)
        claims = json.loads(payload_bytes)
        return AgenticEnvelope.from_dict(claims)


class EnvelopeSigner:
    """
    Signs Agentic Identity Envelopes using an ECDSA P-256 private key.

    Typically you don't need this class directly — use AgentAuthClient.create_envelope()
    which calls the server to sign envelopes. This class is for advanced use cases
    where you need to sign envelopes locally (e.g. offline or edge deployments).

    Example::

        from agentauth import EnvelopeSigner, ToolPermission

        signer = EnvelopeSigner(private_key_pem=PRIVATE_KEY_PEM)
        token = signer.sign(
            agent_id="spiffe://agentauth.io/agent/my-org/abc123",
            declared_intent="Summarize customer emails",
            tool_scope=[ToolPermission("gmail.read", ["read"])],
            tenant_id="my-org",
            ttl_seconds=900,
        )
    """

    def __init__(self, private_key_pem: str, issuer: str = "agentauth") -> None:
        """
        :param private_key_pem: The ECDSA P-256 private key in PEM format.
        :param issuer: The JWT issuer claim value.
        """
        self.private_key_pem = private_key_pem
        self.issuer = issuer
        self._key: Any = None

        if _HAS_PYJWT:
            from cryptography.hazmat.primitives.serialization import load_pem_private_key  # type: ignore[import]
            try:
                self._key = load_pem_private_key(private_key_pem.encode(), password=None)
            except Exception as e:
                raise ValueError(f"Invalid private key PEM: {e}") from e

    def sign(
        self,
        agent_id: str,
        declared_intent: str,
        tool_scope: list[ToolPermission],
        tenant_id: str,
        delegator_id: str = "",
        session_id: str = "",
        ttl_seconds: int = 900,
        delegation_chain: list[str] | None = None,
    ) -> str:
        """
        Sign a new envelope JWT.

        :returns: Signed JWT string.
        :raises RuntimeError: If PyJWT or cryptography is not installed.
        """
        if not _HAS_PYJWT or self._key is None:
            raise RuntimeError(
                "PyJWT and cryptography are required for local signing. "
                "Install them with: pip install agentauth[crypto]"
            )

        import uuid

        now = int(time.time())
        payload: dict[str, Any] = {
            "iss": self.issuer,
            "sub": agent_id,
            "iat": now,
            "exp": now + ttl_seconds,
            "jti": str(uuid.uuid4()),
            "agent_id": agent_id,
            "tenant_id": tenant_id,
            "declared_intent": declared_intent,
            "tool_scope": [ts.to_dict() for ts in tool_scope],
            "ver": "1",
        }
        if delegator_id:
            payload["delegator_id"] = delegator_id
        if session_id:
            payload["session_id"] = session_id
        if delegation_chain:
            payload["delegation_chain"] = delegation_chain

        token = pyjwt.encode(payload, self._key, algorithm="ES256")
        if isinstance(token, bytes):
            return token.decode("utf-8")
        return token


def _matches_pattern(pattern: str, value: str) -> bool:
    """Check if a glob pattern matches a tool name."""
    if pattern == "*" or pattern == value:
        return True
    if pattern.endswith(".*"):
        prefix = pattern[:-2]
        return value == prefix or value.startswith(prefix + ".")
    if pattern.endswith("*"):
        return value.startswith(pattern[:-1])
    return False
